### PR TITLE
Added Twitter Bot Algorith to Tweet from joke.py as Joke_Twitter_Bot.py

### PR DESCRIPTION
**Pull request by Aryainguz**
*Successfully made the algorithm to tweet jokes url from joke.py as a Twitter Bot

*To get credentials of api ,bearertoken, etc visit twitter developer account

FIXES : #22

### CODE REVIEW

**Comment by rohan472000** [comment]
Is it giving link or mene??

**Comment by Aryainguz** [comment]
> Is it giving link or mene??

link

**Comment by rohan472000** [comment]
Ok..give me sometime I'll try to fix it.

**Comment by Aryainguz** [comment]
> Ok..give me sometime I'll try to fix it.

sure

**Comment by rohan472000** [comment]
In the meantime, see my joke.py as I used json to load the meme.

File: Jokes_Twitter_Bot.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+reddit_api_url = "https://www.reddit.com/r/memes/random.json?limit=1"
+
+# Define a user agent to mimic a web browser
+user_agent = (
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
+    "(KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
+)
+
+meme_data = joke.fetch_random_meme(reddit_api_url, user_agent)
+meme_url = joke.extract_meme_url(meme_data)
```
Comment (rohan472000) [inline Jokes_Twitter_Bot.py:14]: After extraction it should load into json or something.

Comment (Aryainguz) [reply Jokes_Twitter_Bot.py:14]: It's only giving back a link of the meme from Reddit

Comment (rohan472000) [reply Jokes_Twitter_Bot.py:14]: That's why I m asking you to load into json then give it to make tweet, see joke.py how i fetched and loaded the data.

I need an hour to look into it deeply as I'm also busy in maintaining other PRs and query, so by that time try something :)

Comment (Aryainguz) [reply Jokes_Twitter_Bot.py:14]: okay sure

**Comment by rohan472000** [comment]
import tweepy
        import requests
        import logging
        import re
        import json
        
        # Constants
        REDDIT_API_URL = "https://www.reddit.com/r/memes/random.json?limit=1"
        USER_AGENT = (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            "(KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
        )
        README_FILE = 'README.md'
        IMAGE_EXTENSIONS_PATTERN = re.compile(r'\.(jpg|jpeg|png)$', re.IGNORECASE)
        
        #credentials
        api_key = "YOUR_API_KEY"
        api_secret = "YOUR_API_SECRET"
        access_token = "YOUR_ACCESS_TOKEN"
        access_token_secret = "YOUR_ACCESS_TOKEN_SECRET"
        
        logging.basicConfig(level=logging.INFO)
        
        def fetch_random_url(api_url, user_agent):
            try:
                response = requests.get(api_url, headers={'User-agent': user_agent})
                response.raise_for_status()
                return response.json()[0]['data']['children'][0]['data']['url']
            except (requests.exceptions.RequestException, json.JSONDecodeError, KeyError, IndexError) as e:
                logging.error(f"An error occurred: {e}")
                return None
        
        def update_readme_with_url(markdown):
            try:
                with open(README_FILE, 'r') as file:
                    contents = file.readlines()
        
                for i, line in enumerate(contents):
                    if "![Funny Meme]" in line:
                        contents[i] = markdown + "\n"
                        break
        
                with open(README_FILE, 'w') as file:
                    file.writelines(contents)
            except (IOError, FileNotFoundError) as e:
                logging.error(f"An error occurred: {e}")
        
        def tweet_meme(meme_url):
            try:
                # Authenticate with Twitter API using Tweepy
                auth = tweepy.OAuthHandler(api_key, api_secret)
                auth.set_access_token(access_token, access_token_secret)
                api = tweepy.API(auth)
        
                # Tweet the meme URL with a hashtag
                tweet_text = f"Check out this meme: {meme_url} #Memes #Jokes #Reddit"
                api.update_status(status=tweet_text)
                logging.info("Tweeted successfully!")
            except tweepy.TweepError as e:
                logging.error(f"Error tweeting: {str(e)}")
        
        def main():
            meme_url = fetch_random_url(REDDIT_API_URL, USER_AGENT)
            if meme_url and IMAGE_EXTENSIONS_PATTERN.search(meme_url):
                markdown = f"Funny Meme"
                update_readme_with_url(markdown)
                tweet_meme(meme_url)
        
        if __name__ == "__main__":
            main()




Try above code with your credentials.

**Comment by Aryainguz** [comment]
> ```
>     import tweepy
>     import requests
>     import logging
>     import re
>     import json
>     
>     # Constants
>     REDDIT_API_URL = "https://www.reddit.com/r/memes/random.json?limit=1"
>     USER_AGENT = (
>         "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
>         "(KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
>     )
>     README_FILE = 'README.md'
>     IMAGE_EXTENSIONS_PATTERN = re.compile(r'\.(jpg|jpeg|png)$', re.IGNORECASE)
>     
>     #credentials
>     api_key = "YOUR_API_KEY"
>     api_secret = "YOUR_API_SECRET"
>     access_token = "YOUR_ACCESS_TOKEN"
>     access_token_secret = "YOUR_ACCESS_TOKEN_SECRET"
>     
>     logging.basicConfig(level=logging.INFO)
>     
>     def fetch_random_url(api_url, user_agent):
>         try:
>             response = requests.get(api_url, headers={'User-agent': user_agent})
>             response.raise_for_status()
>             return response.json()[0]['data']['children'][0]['data']['url']
>         except (requests.exceptions.RequestException, json.JSONDecodeError, KeyError, IndexError) as e:
>             logging.error(f"An error occurred: {e}")
>             return None
>     
>     def update_readme_with_url(markdown):
>         try:
>             with open(README_FILE, 'r') as file:
>                 contents = file.readlines()
>     
>             for i, line in enumerate(contents):
>                 if "![Funny Meme]" in line:
>                     contents[i] = markdown + "\n"
>                     break
>     
>             with open(README_FILE, 'w') as file:
>                 file.writelines(contents)
>         except (IOError, FileNotFoundError) as e:
>             logging.error(f"An error occurred: {e}")
>     
>     def tweet_meme(meme_url):
>         try:
>             # Authenticate with Twitter API using Tweepy
>             auth = tweepy.OAuthHandler(api_key, api_secret)
>             auth.set_access_token(access_token, access_token_secret)
>             api = tweepy.API(auth)
>     
>             # Tweet the meme URL with a hashtag
>             tweet_text = f"Check out this meme: {meme_url} #Memes #Jokes #Reddit"
>             api.update_status(status=tweet_text)
>             logging.info("Tweeted successfully!")
>         except tweepy.TweepError as e:
>             logging.error(f"Error tweeting: {str(e)}")
>     
>     def main():
>         meme_url = fetch_random_url(REDDIT_API_URL, USER_AGENT)
>         if meme_url and IMAGE_EXTENSIONS_PATTERN.search(meme_url):
>             markdown = f"Funny Meme"
>             #update_readme_with_url(markdown)
>             tweet_meme(meme_url)
>     
>     if __name__ == "__main__":
>         main()
> ```
> 
> 
>     
>   
> 
> Try above code with your credentials.

Still only tweets link of meme not image

**Comment by rohan472000** [comment]
import tweepy
    import requests
    import logging
    import re
    import json
    
    # Constants
    REDDIT_API_URL = "https://www.reddit.com/r/memes/random.json?limit=1"
    USER_AGENT = (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
    )
    README_FILE = 'README.md'
    IMAGE_EXTENSIONS_PATTERN = re.compile(r'\.(jpg|jpeg|png)$', re.IGNORECASE)
    
    #credentials
    api_key = "YOUR_API_KEY"
    api_secret = "YOUR_API_SECRET"
    access_token = "YOUR_ACCESS_TOKEN"
    access_token_secret = "YOUR_ACCESS_TOKEN_SECRET"
    
    logging.basicConfig(level=logging.INFO)
    
    def fetch_random_url(api_url, user_agent):
        try:
            response = requests.get(api_url, headers={'User-agent': user_agent})
            response.raise_for_status()
            return response.json()[0]['data']['children'][0]['data']['url']
        except (requests.exceptions.RequestException, json.JSONDecodeError, KeyError, IndexError) as e:
            logging.error(f"An error occurred: {e}")
            return None
    
    def update_readme_with_url(markdown):
        try:
            with open(README_FILE, 'r') as file:
                contents = file.readlines()
    
            for i, line in enumerate(contents):
                if "![Funny Meme]" in line:
                    contents[i] = markdown + "\n"
                    break
    
            with open(README_FILE, 'w') as file:
                file.writelines(contents)
        except (IOError, FileNotFoundError) as e:
            logging.error(f"An error occurred: {e}")
    
    def tweet_meme(meme_url):
        try:
            # Authenticate with Twitter API using Tweepy
            auth = tweepy.OAuthHandler(api_key, api_secret)
            auth.set_access_token(access_token, access_token_secret)
            api = tweepy.API(auth)
    
            # Tweet the meme URL with a hashtag
            tweet_text = f"Check out this meme: {meme_url} #Memes #Jokes #Reddit"
            api.update_status(status=tweet_text)
            logging.info("Tweeted successfully!")
        except tweepy.TweepError as e:
            logging.error(f"Error tweeting: {str(e)}")
    
    def main():
        meme_url = fetch_random_url(REDDIT_API_URL, USER_AGENT)
        if meme_url and IMAGE_EXTENSIONS_PATTERN.search(meme_url):
            markdown = f"Funny Meme"
            #update_readme_with_url(markdown)
            tweet_meme(markdown )
    
    if __name__ == "__main__":
        main()






try this one, i just changed : ` tweet_meme(markdown )`

**Comment by Aryainguz** [comment]
The above modules are only included in paid twitter api

**Comment by rohan472000** [comment]
> The above modules are only included in paid twitter api

But previously it was free.

**Comment by Aryainguz** [comment]
> > The above modules are only included in paid twitter api
> 
> But previously it was free.

yes I could not check it as I am using free api version

**Comment by rohan472000** [comment]
Anyway, try if you can to manage those api.

File: Jokes_Twitter_Bot.py (hunk 1 above)
Comment (ShubhamNagure) [inline Jokes_Twitter_Bot.py:14]: Where is this **extract_meme_url** method? I didn't find it in **joke.py**; is it in some specific branch?

Comment (rohan472000) [reply Jokes_Twitter_Bot.py:14]: It is there..look carefully

Comment (rohan472000) [reply Jokes_Twitter_Bot.py:14]: Ahh... previously it was there but one contributor made it in less lines of code by removing those methods.

**Pull request closed without merging**